### PR TITLE
BN-124|BN-175  New patient registration page with basic details and Additional Details

### PR DESCRIPTION
**Pull request by devsingh05298**
**JIRA** → [BN-124](https://bahmni.atlassian.net/browse/BN-124), [BN-175](https://bahmni.atlassian.net/browse/BN-175)

### **Description**
This PR introduces a comprehensive patient registration feature that enables healthcare staff to create new patient records in the Bahmni system. The implementation includes a multi-section form with robust validation, address hierarchy integration, and real-time age/DOB calculations.

### 1. __Patient Registration Form__

- Multi-section form layout with Basic Information, Address Information, Contact Information, and Additional Information sections
- Photo upload and capture functionality placeholders
- Patient identifier format selection with dynamic prefix management
- Manual entry toggle for identifier creation

### 2. __Form Validations__

- __Name Validation__: Ensures only alphabetic characters and spaces are allowed in name fields

- __Required Field Validation__: First name, last name, gender, and date of birth are mandatory

- __Age Validation__:

  - Years: Maximum 120
  - Months: Maximum 11
  - Days: Maximum 31

- __Phone Number Validation__: Accepts only numeric input

- __Address Validation__: Enforces selection from dropdown suggestions for district, state, and pincode

### 3. __Address Hierarchy Integration__

- Real-time address suggestions with auto-complete functionality
- Hierarchical auto-population: When pincode is selected, district and state auto-populate
- Debounced API calls (300ms) for optimal performance
- Minimum 2 characters required to trigger suggestions
- Visual feedback with parent address information in suggestions

### 4. __Age/DOB Bidirectional Calculation__

- Automatic DOB calculation when age is entered (years/months/days)
- Automatic age calculation when DOB is selected
- "Estimated" checkbox automatically checked when age is entered manually
- Supports entry in years, months, and days for precise age specification

### 5. __Gender Selection__

- Dynamic gender options fetched from global properties API
- Support for Male, Female, and Other gender options
- Internationalized gender labels

### UI Components

__Core Components Used:__

- `BaseLayout` - Main layout wrapper with header
- `Header` - Custom header component with breadcrumb navigation
- `TextInput` - For name, contact, and address fields
- `Dropdown` - For gender and patient ID format selection
- `DatePicker` & `DatePickerInput` - For date of birth selection
- `Checkbox` - For manual entry and estimated DOB
- `Button` - For form actions (Save, Print, Start OPD Visit)
- `Tile` - For section grouping

__Custom Styling:__

- Module-based SCSS for component-specific styles
- Responsive grid layout for form fields
- Address suggestion dropdown with hierarchical display
- Error message styling for inline validation

### State Management

__Local State (useState):__

- `formData` - Contains all form field values (23 fields)
- `nameErrors` - Validation errors for name inputs
- `validationErrors` - Required field validation errors
- `ageErrors` - Age-specific validation errors
- `addressErrors` - Address field validation errors
- `suggestions` - Address hierarchy suggestions for district, state, pincode
- `showSuggestions` - Visibility state for suggestion dropdowns
- `addressSelectedFromDropdown` - Tracks if address fields were selected vs manually typed
- `dobEstimated` - Tracks if DOB is estimated or exact

__Computed State (useMemo):__

- `identifierPrefixes` - Memoized identifier prefixes from API
- `identifierSources` - Memoized source UUIDs by prefix
- `primaryIdentifierType` - Memoized primary identifier type UUID
- `genders` - Memoized and translated gender options

### Custom Hooks

__useQuery (TanStack Query):__

- `identifierData` - Fetches all identifier type data (prefixes, sources, primary type)

  - 5-minute stale time
  - 10-minute cache time
  - Retry logic with exponential backoff (3 attempts, max 30s delay)

- `gendersFromApi` - Fetches gender options from global property

__useMutation (TanStack Query):__

- `createPatientMutation` - Handles patient creation with success/error callbacks

  - Success: Shows notification and navigates to patient details
  - Error: Shows error notification

__Custom Callbacks (useCallback):__

- `handleInputChange` - Generic input change handler
- `handleNameChange` - Name input with alphabetic validation
- `handlePhoneChange` - Phone input with numeric validation
- `handleDateOfBirthChange` - DOB change with age calculation
- `handleAgeChange` - Age change with DOB calculation and validation
- `debouncedSearchAddress` - Debounced address hierarchy search
- `handleAddressInputChange` - Address input with suggestion trigger
- `handleSuggestionSelect` - Address suggestion selection with auto-population

### API Integration

__Services Used:__

1. `getIdentifierData()` - Fetches identifier prefixes, sources, and primary identifier type
2. `getGenders()` - Fetches gender options from global properties
3. `getAddressHierarchyEntries(addressField, searchText)` - Fetches address suggestions
4. `createPatient(patientRequest)` - Creates new patient record

__API Models:__

- `CreatePatientRequest` - Patient creation request structure
- `PatientAttribute` - Patient attribute (phone, email) structure
- `PatientAddress` - Address structure
- `AddressHierarchyEntry` - Address hierarchy entry with parent references

### Performance & Optimization

__Query Optimization:__

- Stale time configured for reducing unnecessary API calls
- Cache time configured for optimal memory usage
- Retry logic with exponential backoff for network resilience
- Parallel API calls for initial data loading

__Debouncing:__

- 300ms debounce on address search to reduce API calls
- Minimum 2 characters required before triggering search

__Memoization:__

- `useMemo` for expensive computations (identifier processing, gender mapping)
- `useCallback` for stable function references preventing unnecessary re-renders

### Internationalization

__Translation Keys Added:__ 50+ new translation keys covering:

- All form labels and placeholders
- Validation error messages
- Button text and actions
- Section titles and breadcrumbs
- Gender options

__i18n Files Updated:__

- `distro/src/assets/locales/locale_en.json` - English translations
- `apps/registration/public/locales/locale_en.json` - Registration app translations
- `distro/src/assets/locales/locale_es.json` - Spanish translations (if applicable)

__Translation Categories:__

- `CREATE_PATIENT_*` - Form-related translations
- `CREATE_PATIENT_VALIDATION_*` - Validation error messages
- `CREATE_PATIENT_SECTION_*` - Section titles
- `CREATE_PATIENT_GENDER_*` - Gender options

### Accessibility

__Keyboard Navigation:__

- All form fields are keyboard accessible
- Tab order follows logical flow
- Dropdown suggestions navigable via keyboard

__ARIA Labels:__

- Form inputs have proper labels
- Error messages are associated with inputs
- Required fields marked with aria-required

__Screen Reader Support:__

- Semantic HTML structure
- Proper heading hierarchy
- Error messages announced
- Button text descriptive

## Limitations

1. __Photo Upload/Capture__: Currently placeholders - functionality not implemented
2. __Print Registration Card__: Button present but functionality not implemented
3. __Start OPD Visit__: Button present but functionality pending integration
4. __Birth Time__: Field present but not included in patient creation payload
5. __Patient Attributes__: Phone number and email attribute UUIDs are hardcoded placeholders
6. __Address Hierarchy__: Only district, state, and pincode support suggestions; other fields are free text
7. __Back to Search__: Button present but action not implemented

### Unit Tests

__Test File:__ `apps/registration/src/pages/createPatientPage/__tests__/createPatient.test.tsx`

__Testing Utilities:__

- React Testing Library for component testing
- Jest for test runner and assertions
- User Event library for realistic user interactions
- Jest-axe for accessibility testing
- Mock implementations for APIs and routing

---

__Related Issues:__ BN-124, BN-175 (Address and contact information validation)

### **Screenshots**

| Feature / State | Screenshot |
|------------------|------------|
| Create New Patient | Screenshot 2025-10-10 at 3 16 35 PM |

_Note: Please ensure sensitive information is not included in screenshots._

---

### **Next Steps**
 
 

---

> [!IMPORTANT]
> **Checklist**
>
> - [ ] Code adheres to project linting and formatting standards.
> - [ ] New components added to Storybook.
> - [ ] Tests written and passing (unit + integration).
> - [ ] Labels and text support i18n.
> - [ ] Follows accessibility and responsive design guidelines.
> - [ ] PR reviewed by at least one other developer.

---

### **Reviewer(s)**

@bahnew/developers 
_Kindly review the proposed changes when convenient. Your feedback is appreciated._

---

### CODE REVIEW

File: apps/registration/src/components/Header/Header.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import styles from './styles/index.module.scss';
+
+export interface BreadcrumbItem {
+  label: string;
+  href?: string;
+  onClick?: () => void;
+}
+
+export interface HeaderProps {
+  breadcrumbs: BreadcrumbItem[];
```
Comment (arshiyaTW2021) [inline apps/registration/src/components/Header/Header.tsx]: Instead of custom implementation could have imported  Breadcrumb, BreadcrumbItem from '@carbon/react',  add it to bahmni-design-system and export it to maintain consistency and maintainabilty

File: apps/registration/src/utils/ageUtils.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    };
+
+    let years = to.y - from.y;
+    let months = to.m - from.m;
+    let days = to.d - from.d;
+
+    // Adjust negative days
+    if (days < 0) {
+      months -= 1;
+      const prevMonthLastDay = new Date(to.y, to.m, 0).getDate(); // last day of previous month
```
Comment (arshiyaTW2021) [inline apps/registration/src/utils/ageUtils.ts]: Are we handling edge cases like leap years properly over here? Also, have we written a test case for the same? 
For Negative Days, wrong reference month, leap years and for time zone issues, it is going to give buggy results
 All of these can be avoided by using date-fns. 
 
1. **Negative Days Not Handled Properly**
if (days < 0) {
  months -= 1;
  const prevMonthLastDay = new Date(to.y, to.m, 0).getDate();
  days += prevMonthLastDay;
}
Problem: Uses if instead of while, so it only runs once. If days are still negative after one adjustment, it returns negative days! Example:
Born: Jan 31, 2024
Today: March 1, 2024
Buggy result: { years: 0, months: 1, days: -1 } 
Correct result: { years: 0, months: 1, days: 1 } 
Why it happens:
Initial: days = 1 - 31 = -30
Borrows from month: days = -30 + 29 (Feb days) = -1
Still negative, but code doesn't loop again!

2. **Wrong Reference Month for Day Calculation**
Example:
Born: May 31, 2024
Today: July 30, 2024
Buggy result: { years: 0, months: 1, days: 29 } 
Correct result: { years: 0, months: 1, days: 30 } 
Why it happens:
Initial: days = 30 - 31 = -1
Code does: new Date(2024, 6, 0).getDate() → Gets June (30 days)
days = -1 + 30 = 29
But May 31 → July 30 is actually 60 days = 1 month (30 days) + 30 days

3. **Leap Year Edge Cases**
Example:
Born: Feb 29, 2020 (leap year)
Today: March 1, 2021
Buggy result: { years: 1, months: 0, days: 0 } 
Correct result: { years: 1, months: 0, days: 3 } 
The algorithm doesn't handle leap year births correctly when crossing non-leap years.

4. **Timezone Issues**
Problem: When parsing ISO date strings without time (like "2024-01-15"), JavaScript interprets them as UTC midnight. For users in negative UTC timezones (e.g., America/Los_Angeles UTC-8), calling .getDate() on this returns the previous day in local time. Example:
const dob = new Date("2024-01-15"); // User enters this
// In Los Angeles (UTC-8): This becomes Jan 14, 4:00 PM PST
dob.getDate(); // Returns 14 (not 15!) in LA timezone

**All of this can be avoided and will save a lot of redundant code by using date-fns** 

`import { intervalToDuration, formatISO, sub } from 'date-fns';

export const diffInYearsMonthsDays = (dateFrom: Date, dateTo: Date = new Date()) => {
  return intervalToDuration({ start: dateFrom, end: dateTo });
};

export const calculateBirthDate = (age: Age): string => {
  const dob = sub(new Date(), {
    years: age.years || 0,
    months: age.months || 0,
    days: age.days || 0,
  });
  return formatISO(dob, { representation: 'date' }); // yyyy-MM-dd
};`

Have you tried this approach already?

File: apps/registration/src/utils/ageUtils.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ * Convert Date object to ISO string (yyyy-mm-dd)
+ */
+export const formatToISO = (date: Date): string => {
+  return date.toISOString().split('T')[0];
+};
+
+/**
+ * Parse display string (dd/mm/yyyy) or ISO string (yyyy-mm-dd) to Date object
+ */
+export const parseDateStringToDate = (s: string): Date | null => {
```
Comment (arshiyaTW2021) [inline apps/registration/src/utils/ageUtils.ts]: Again, can this customised function be avoided by utilising date-fns perhaps?

File: packages/bahmni-services/src/patientService/patientService.ts [hunk 4]
```diff
@@ -1,14 +1,28 @@
...
+  FormattedPatientData,
+  PatientSearchResultBundle,
+  IdentifierTypesResponse,
+  AppSettingsResponse,
+  CreatePatientRequest,
+  CreatePatientResponse,
+  AddressHierarchyEntry,
+} from './models';
 
 export const getPatientById = async (patientUUID: string): Promise<Patient> => {
```
Comment (arshiyaTW2021) [inline packages/bahmni-services/src/patientService/patientService.ts]: have we handled the scenarios when patientUUId is empty or invalid? Is there any validation for the same or is that being handled differently and won't affect our current scope of the story?

File: distro/src/assets/locales/locale_es.json [hunk 5]
```diff
@@ -83,6 +83,68 @@
...
+  "CREATE_PATIENT_START_OPD_VISIT": "Iniciar visita OPD",
+  "CREATE_PATIENT_VALIDATION_NAME_INVALID": "No se permiten números ni caracteres especiales",
+  "CREATE_PATIENT_VALIDATION_FIRST_NAME_REQUIRED": "El nombre es obligatorio",
+  "CREATE_PATIENT_VALIDATION_LAST_NAME_REQUIRED": "El apellido es obligatorio",
+  "CREATE_PATIENT_VALIDATION_GENDER_REQUIRED": "El género es obligatorio",
+  "CREATE_PATIENT_VALIDATION_DATE_OF_BIRTH_REQUIRED": "La fecha de nacimiento es obligatoria",
+  "CREATE_PATIENT_VALIDATION_SELECT_FROM_DROPDOWN": "Seleccione la entrada del menú desplegable",
+  "CREATE_PATIENT_VALIDATION_AGE_YEARS_MAX": "La edad no puede ser mayor de 120 años",
+  "CREATE_PATIENT_VALIDATION_AGE_MONTHS_MAX": "Los meses no pueden ser más de 11",
+  "CREATE_PATIENT_VALIDATION_AGE_DAYS_MAX": "Los días no pueden ser más de 31",
```
[the remark this answers is not in the document]
Comment (ROHIT-V-TW) [reply distro/src/assets/locales/locale_es.json]: In locale_es line 112 there’s Age_Placeholder which is not needed hence the extra key (removed it)

File: apps/registration/src/utils/ageUtils.ts (hunk 3 above)
Comment (ROHIT-V-TW) [reply apps/registration/src/utils/ageUtils.ts]: Used date-fns as mentioned in another conversation

File: apps/registration/src/components/Header/Header.tsx [hunk 6]
```diff
@@ -31,49 +32,18 @@ const Header: React.FC<HeaderProps> = ({
...
-                    }
-                  >
-                    {item.label}
-                  </span>
-                )}
-              </React.Fragment>
-            ))}
-          </nav>
-        </div>
+        <Breadcrumb noTrailingSlash className={styles.carbonBreadcrumb}>
```
Comment (arshiyaTW2021) [inline apps/registration/src/components/Header/Header.tsx]: We are supposed to use design components fro bahmni-design-system . If they are not there already, they are to be added there and then used from there. In atoms folder, this could have been added and utilised from there. Think of it this way, moving further we may require it at other places too.

**Review by arshiyaTW2021 — APPROVED** [review]
